lib/lcevc/lcevc_dil.js

```diff
@@ -0,0 +1,281 @@
+
+    /** @private {?HTMLElement} */
+    this.lcevcContainer_ = null;
+
+    /** @private {?HTMLElement} */
+    this.videoControls_ = null;
+
+    /** @private {ResizeObserver} */
+    this.resizeObserver_ = null;
+
+    this.level_ = -1;
+    this.containerFormat_ = 0;
+    this.media_ = /** @type {HTMLVideoElement} */ (media);
+    this.canvas_ = /** @type {HTMLCanvasElement} */ (canvas);
+    this.dilConfig_ = dilConfig;
+
+    // This line is optional forcing the config to hide the LCEVC watermark
+    // that is displayed on the left hand side top corner.
+    if (!this.dilConfig_['logo']) {
+      this.dilConfig_['logo'] = false;
+    }
+  }
+
+  /**
+   * Append data to the LCEVC Dil.
+   * @param {BufferSource} data
+   */
+  appendBuffer(data) {
+    if (this.dil_) {
+      this.dil_.appendBuffer(data, 'video', this.level_);
+    }
+  }
+
+  /**
+  * Hide the canvas and show original video element
+  */
+  hideCanvas() {
+    if (this.dil_) {
+      this.media_.style.display = 'block';
+      this.canvas_.style.display = 'none';
+    }
+  }
+
+  /**
+    * Show the canvas and hide original video element
+    */
+  showCanvas() {
+    if (this.dil_) {
+      this.media_.style.display = 'none';
+      this.canvas_.style.display = 'block';
+    }
+  }
+
+  /**
+   * Create LCEVC Dil.
+   * @private
+   */
+  createDil_() {
+    if (this.isSupported() && !this.dil_) {
+      if (!this.canvas_) {
+        this.createCanvas_();
+      }
+      if (LcevcDil.SupportObject.webGLSupport(this.canvas_)) {
+        this.dil_ = new LcevcDil.LcevcDil(
+            this.media_,
+            this.canvas_,
+            this.dilConfig_);
+
+        this.setControlsStyle_();
+      }
+    }
+  }
+
+  /**
+   * Reset LCEVC Dil.
+   */
+  async reset() {
+    const INVALID_LEVEL = -2;
+    if (this.dil_) {
+      this.dil_.setCurrentLevel(INVALID_LEVEL);
+      await Promise.resolve();
+      this.dil_.setLevelSwitching(this.level_, 1);
+    }
+  }
+
+  /**
+   * Close LCEVC Dil.
+   */
+  close() {
+    if (this.dil_) {
+      this.dil_.close();
+      this.dil_ = null;
+    }
+    this.lcevcContainer_.remove();
+    if (this.resizeObserver_) {
+      this.resizeObserver_.disconnect();
+      this.resizeObserver_ = null;
+    }
+    if (this.canvas_) {
+      this.canvas_.remove();
+    }
+  }
+
+  /**
+   * Check if LCEVC Dil exists.
+   */
+  isSupported() {
+    if (typeof libDPIModule === 'undefined') {
+      shaka.log.alwaysWarn(
+          'Could not Find LCEVC Library dependencies on this page');
+    }
+
+    if (typeof LcevcDil === 'undefined') {
+      shaka.log.alwaysWarn('Could not Find LCEVC Library on this page');
+    } else {
+      if (!LcevcDil.SupportStatus) {
+        shaka.log.alwaysWarn(LcevcDil.SupportError);
+      }
+    }
+
+    return typeof LcevcDil !== 'undefined' &&
+      typeof libDPIModule !== 'undefined' &&
+      LcevcDil.SupportStatus;
+  }
+
+  /**
+   * Create canvas for LCEVC Dil to render if not exists.
```

Comment:
As a matter of policy, we don't create DOM elements at this level.  Only the UI can do that.

Reply:
Is there any specific workflow that needs to be followed to create DOM elements? can you please help me point to a reference. Thanks.

Reply:
If I could propose an alternative design, perhaps...

You could pass in the canvas, not in the player constructor, but in some setter method. `setLCEVCCanvas` or something. And the UI library can have a similar method of its own.

Then make the LCEVC code not make a canvas at all, and the UI library can make the default canvas if the user doesn't provide a custom on.

If the user does not use the UI library and also doesn't provide a canvas, you could throw an error or something.

Admittedly that would also require some parts of the resize logic to be exposed to the UI, so it can know the aspect ratio of the DIL and such, which seems awkward design-wise...

Reply:
Shaka runs in two versions:
 1. with UI (UI layer creates, positions DOM elements, creates Player object)
 2. without UI (app creates, positions DOM elements, provides them to Player)

What @theodab is proposing is more in line with that philosophy.

There are some gray areas, though.  Text rendering happens within an app-provided or UI-provided div.  Within that div, the text rendering plugin creates and owns its own DOM hierarchy.  But the top-level item (where all the text should overlay the video) is owned by the UI or by the app (in the non-Shaka-UI case).

Here, since canvas is the only element, I feel like the app or the UI should provide the canvas, aligned to the video element and positioned on top of it in Z order.

Reply:
Please drop createCanvas_() and move this functionality to the UI layer.  See my new comments above at the createCanvas_() call site.